app/PaymentDrivers/CheckoutCom/Utilities.php

```diff
@@ -39,4 +46,131 @@ public function convertToCheckoutAmount($amount, $currency)
         // https://docs.checkout.com/resources/calculating-the-value#Calculatingthevalue-Option3:Thevaluedividedby100valuediv100
         return round($amount * 100);
     }
+
+    private function processSuccessfulPayment(\Checkout\Models\Payments\Payment $_payment)
+    {
+        if ($this->payment_hash->data->store_card) {
+            // $this->saveCreditCard();
+        }
+
+        $data = [
+            'payment_method' => $_payment->source['id'],
+            'payment_type' => PaymentType::parseCardType(strtolower($_payment->source['scheme'])),
+            'amount' => $this->payment_hash->data->value,
+        ];
+
+        $payment = $this->checkout->createPayment($data, \App\Models\Payment::STATUS_COMPLETED);
+
+        $this->payment_hash->payment_id = $payment->id;
+        $this->payment_hash->save();
+
+        $this->checkout->attachInvoices($payment, $this->payment_hash);
+
+        $payment->service()->updateInvoicePayment($this->payment_hash);
+
+        event(new PaymentWasCreated($payment, $payment->company, Ninja::eventVars()));
+
+        SystemLogger::dispatch(
+            ['response' => $_payment, 'data' => $data],
+            SystemLog::CATEGORY_GATEWAY_RESPONSE,
+            SystemLog::EVENT_GATEWAY_SUCCESS,
+            SystemLog::TYPE_CHECKOUT,
+            $this->checkout->client
+        );
+
+        return redirect()->route('client.payments.show', ['payment' => $this->checkout->encodePrimaryKey($payment->id)]);
+    }
+
+    public function processUnsuccessfulPayment(\Checkout\Models\Payments\Payment $_payment)
+    {
+        PaymentFailureMailer::dispatch(
+            $this->checkout->client,
+            $_payment,
+            $this->checkout->client->company,
+            $this->payment_hash->data->value
+        );
+
+        $message = [
+            'server_response' => $_payment,
+            'data' => $this->payment_hash->data,
+        ];
+
+        SystemLogger::dispatch(
+            $message,
+            SystemLog::CATEGORY_GATEWAY_RESPONSE,
+            SystemLog::EVENT_GATEWAY_FAILURE,
+            SystemLog::TYPE_CHECKOUT,
+            $this->checkout->client
+        );
+
```

Comment:
Abstract the failures to a generic \Exception which has a render() method which displays an appropriate page to the end user... this should be handled in the base driver.

Reply:
Fixed here: https://github.com/invoiceninja/invoiceninja/pull/4202/commits/6e6e73825c1f1315f1aad9a3460e1cdf5f09f648